compiler/rustc_abi/src/lib.rs

```diff
@@ -1489,6 +1489,29 @@ pub enum Variants<FieldIdx: Idx, VariantIdx: Idx> {
     },
 }
 
+impl<FieldIdx: Idx, VariantIdx: Idx> Variants<FieldIdx, VariantIdx> {
+    // Returns all used niche values.
```

Comment:
Yes, it iterates the discriminant and returns the corresponding niche.

Reply:
But this is returning a single value per variant, and niches are at least a range. And this is based on the discriminant start, so I'm pretty sure this just returns discriminants.

Then we fold these using `restrict_for_offset`? The values we're passing to that function aren't offsets. Does that function just so happen to implement the arithmetic that we need?

I think I'm going to try to work out this with pencil and paper, because I feel like the math works out in your implementation, but I still don't think I can see how.

Reply:
IIUC, a variant corresponds to a niche. I convert to a discriminant(`variant_idx`) to a niche by `(d - niche_variants.start).wrapping_add(niche_start)`, so this returns an iterator for all niches.

Then we pass a niche as a pointer, the max alignment of a niche is `max_for_offset(niche)`. For example:
- the max alignment of `8` is `8`
- the max alignment of `24`(`0b11000`) is `8`
- the max alignment of `26`(`0b11010`) is `2`

In order for the alignment to be legal, we must choose the min alignment and `restrict_for_offset` is I want.